### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by TalMekler**

### CODE REVIEW

File: scripts/Item.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    setDesc(desc) {
+        this.desc = desc
+    }
+    setPrice(price) {
+        this.price = price
+    }
+
+    makeHtmlItem() {
+        const st = "'" +this.img + "', " + this.title + ", " + this.desc + ", " + this.price + ", " + this.id
```
Comment (danielmargolin) [inline scripts/Item.js:38]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

Comment (danielmargolin) [reply scripts/Item.js:38]: Plus, can't find the usage of `st`

File: scripts/Item.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        return "<div class='item' attr_img='"+ this.img +"' attr_title='" + this.title + "' attr_desc='" + this.desc + "' attr_price='" + this.price + "' attr_id='" + this.id +"'> <div class='item__img'><img src='" + this.img + "' alt='item img'></div><div class='item-details'><p class='item__title'>" + this.title + "</p><p class='item__desc'>" + this.desc + "</p><p class='item__price'><span>Price: </span>" + this.price + "₪</p> <div class='add-to-cart-wrapper'><button class='add-to-cart-btn'>Add</button></div></div></div>"
```
Comment (danielmargolin) [inline scripts/Item.js]: Instead of of creating an unreadable html string, switch to `createElement`

File: scripts/Cart.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        this.cartlines.forEach(cl => {
+            s += parseInt(cl.getTotalPrice())
+        })
+        return s
+    }
+
+    isCartlineExist(cartline) {
+        console.log("isCartlineExist, wanted cartline id = " + cartline.getID());
+        
+        for(let i = 0; i< this.cartlines.length; i++) {
```
Comment (danielmargolin) [inline scripts/Cart.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

Comment (danielmargolin) [reply scripts/Cart.js]: will it work instead?

File: scripts/Cartline.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                            <h4><span class="qtt">1</span>x title</h4>
+                            <p class="desc">desc</p>
+                        </div>
+                    </div>
+                    <p class="total-price"><span>500</span>₪</p>
+                </div> 
+*/
+
+    makeHtmlElement() {
+        return "<div class='cartline' attr-id='" + this.id + "'> <div class='delete-cartitem'><i class='fa-solid fa-xmark'></i></div><div class='item-details-wrapper'><div class='item-img'><img src='" + this.img + "' alt='"+ this.title +"'></div><div class='item-text'><h4><span class='qtt'>"+ this.qtt +"</span>x " + this.title + "</h4><p class='desc'>" + this.desc + "</p></div></div><p class='total-price'><span>" + this.getTotalPrice() + "</span>₪</p></div>"
```
Comment (danielmargolin) [inline scripts/Cartline.js]: `createElement`

File: scripts/script.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
+import Cartline from "./Cartline";
+import Cart from "./Cart";
+import Catalog from "./Catalog";
+import Item from "./Item";
+import data from "../data.json";
+
+document.querySelector(".open-close-cart").addEventListener("click", () => {
+  document.querySelector(".cart-wrapper").classList.toggle("active");
```
Comment (danielmargolin) [inline scripts/script.js:9]: Create a function called `handleCartClick`?

File: scripts/Item.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        this.title = title
+    }
+    setDesc(desc) {
+        this.desc = desc
+    }
+    setPrice(price) {
+        this.price = price
+    }
+
+    makeHtmlItem() {
```
Comment (danielmargolin) [inline scripts/Item.js]: Be consistent: `makeHtmlElement` or `getHtml`

File: scripts/script.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+//   if(maxHeightImg > 300)
+//     maxHeightImg =300
+document
+  .querySelectorAll(".catalog .item .item__img")
+  .forEach((itemImgElement) => {
+    itemImgElement.style.height = maxHeightImg + "px";
+  });
+
+console.log(maxHeightImg);
```
Comment (danielmargolin) [inline scripts/script.js]: redundant console.log

File: scripts/Cartline.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        return this.qtt
+    }
+    setQtt(qtt){
+        this.qtt = qtt
+    }
+    getTotalPrice() {
+        return this.qtt * this.pricePerOne
+    }
+
+/*
```
Comment (danielmargolin) [inline scripts/Cartline.js]: remove comment

File: scripts/Cart.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        this.makeHtmlElement()
+    }
+
+    removeCartline(id) {
+        console.log("getCartLine"+this.getCartlineByID(id));
+        if(this.getCartlineByID(id) != null){
+            console.log("inside");
+            const cl = this.getCartlineByID(id)
+            let i;
+            for(i = 0; i < this.cartlines.length; i++) {
```
Comment (danielmargolin) [inline scripts/Cart.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

File: scripts/Cart.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            this.setNumberOfItems(this.getNumberOfItem + 1)
+        }
+        console.log("cart: " + this.cartlines);
+        this.makeHtmlElement()
+    }
+
+    removeCartline(id) {
+        console.log("getCartLine"+this.getCartlineByID(id));
+        if(this.getCartlineByID(id) != null){
+            console.log("inside");
```
Comment (danielmargolin) [inline scripts/Cart.js]: remove console.log

File: scripts/Cart.js [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        }else {
+            this.cartlines.push(new Cartline(cartline.getImg(), cartline.getTitle(), cartline.getDesc(), cartline.getPricePerOne(), cartline.getID()))
+            this.setNumberOfItems(this.getNumberOfItem + 1)
+        }
+        console.log("cart: " + this.cartlines);
+        this.makeHtmlElement()
+    }
+
+    removeCartline(id) {
+        console.log("getCartLine"+this.getCartlineByID(id));
```
Comment (danielmargolin) [inline scripts/Cart.js]: remove console.log

File: scripts/Cart.js [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        return null
+    }
+
+    addCartline(cartline) {
+        if(this.isCartlineExist(cartline)){
+            console.log("TRUE!");
+            const cl = this.getCartlineByID(cartline.getID())
+            cl.setQtt(cl.getQtt() + 1)
+            this.setNumberOfItems(this.getNumberOfItem + 1)
+            console.log(this.cartlines);
```
Comment (danielmargolin) [inline scripts/Cart.js]: remove console.log

File: scripts/Cart.js [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+            if(this.cartlines[i].getID() == id){
+                return this.cartlines[i];
+            }
+        }
+        return null
+    }
+
+    addCartline(cartline) {
+        if(this.isCartlineExist(cartline)){
+            console.log("TRUE!");
```
Comment (danielmargolin) [inline scripts/Cart.js]: remove console.log

File: scripts/Cart.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+            const res = (this.cartlines[i].getID() == cartline.getID());
+            if(res == true){
+                console.log("res: " + (res)); 
+                return true;
+            }
+        }
+        return false;
+    }
+
+    getCartlineByID(id) {
```
Comment (danielmargolin) [inline scripts/Cart.js]:
```
getCartLineById (id) {
   return this.cartLines.find(cartLine => cartLine.getId() === id)
} 
```

File: scripts/Cart.js [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+        this.totalPrice = 0
+    }
+
+    getCartlines() {
+        return this.cartlines
+    }
+    getNumberOfItem() {
+        return this.numberOfItems
+    }
+    getTotalPrice() {
```
Comment (danielmargolin) [inline scripts/Cart.js]: A `get...` function that manipulates a property? Careful!!!

File: scripts/Cart.js [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+        return s
+    }
+
+    isCartlineExist(cartline) {
+        console.log("isCartlineExist, wanted cartline id = " + cartline.getID());
+        
+        for(let i = 0; i< this.cartlines.length; i++) {
+            const res = (this.cartlines[i].getID() == cartline.getID());
+            if(res == true){
+                console.log("res: " + (res));
```
Comment (danielmargolin) [inline scripts/Cart.js]: console.log

File: scripts/Cart.js [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+                document.querySelector(".cart-wrapper").classList.toggle("active");
+                document.querySelector(".catalog").classList.toggle("active");
+            }
+        }
+        this.makeHtmlElement()
+        
+        // console.log(this.cartlines);
+    }
+
+    makeHtmlElement() {
```
Comment (danielmargolin) [inline scripts/Cart.js]: Be consistent! This `makeHtmlElement` renders to the DOM (unlike other `makeHtmlElement` functions in other classes that just return html.

File: scripts/Cart.js [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+        const cartBox = document.querySelector(".cart-box");
+        cartBox.innerHTML = ""
+        this.cartlines.forEach(cl => {
+            cartBox.innerHTML += cl.makeHtmlElement()
```
Comment (danielmargolin) [inline scripts/Cart.js]: Don't append to the DOM in iteration (in some cases can cause performance issues)!
store it in a string and append it once.

File: scripts/script.js [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+  document.querySelector(".cart-wrapper").classList.toggle("active");
+  document.querySelector(".catalog").classList.toggle("active");
+});
+
+document.querySelector("main").style.height =
+  window.innerHeight - document.querySelector("header").offsetHeight + "px";
+
+const catalog = new Catalog();
+data.items.map((item, i) => {
+  const ITEM = new Item(item.img, item.title, item.desc, item.price, i);
```
Comment (danielmargolin) [inline scripts/script.js:22]: why uppercase?

File: scripts/script.js [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+
+let maxHeightImg = 0;
+document
+  .querySelectorAll(".catalog .item .item__img")
+  .forEach((itemImgElement) => {
+    if (maxHeightImg < itemImgElement.offsetHeight)
+      maxHeightImg = itemImgElement.offsetHeight;
+  });
+
+//   if(maxHeightImg > 300)
```
Comment (danielmargolin) [inline scripts/script.js]: remove comment

File: scripts/script.js [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+const catalog = new Catalog();
+data.items.map((item, i) => {
+  const ITEM = new Item(item.img, item.title, item.desc, item.price, i);
+  catalog.addItem(ITEM);
+});
+
+let maxHeightImg = 0;
+document
+  .querySelectorAll(".catalog .item .item__img")
+  .forEach((itemImgElement) => {
```
Comment (danielmargolin) [inline scripts/script.js:29]: reduce?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce

File: scripts/script.js [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+    itemImgElement.style.height = maxHeightImg + "px";
+  });
+
+console.log(maxHeightImg);
+
+const searchBar = document.querySelector(".search-bar-wrapper input");
+searchBar.addEventListener("input", () => {
+  console.log("search bar changed");
```
Comment (danielmargolin) [inline scripts/script.js]: remove console.log

File: scripts/script.js [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+  document.querySelectorAll(".item").forEach((itemElement) => {
+    console.log(itemElement);
```
Comment (danielmargolin) [inline scripts/script.js]: console.log

File: scripts/script.js [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+    if (
+      itemElement
+        .querySelector(".item__title")
+        .innerHTML.toLocaleLowerCase()
+        .includes(searchBar.value.toLocaleLowerCase())
+    )
+      itemElement.classList.remove("unvisible");
+    else itemElement.classList.add("unvisible");
```
Comment (danielmargolin) [inline scripts/script.js]: else {
...
}

File: scripts/script.js [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+document.querySelectorAll(".catalog .item").forEach((item) => {
+  item.querySelector(".add-to-cart-btn").addEventListener("click", () => {
+    const newCL = new Cartline(
+      item.getAttribute("attr_img"),
+      item.getAttribute("attr_title"),
+      item.getAttribute("attr_desc"),
+      item.getAttribute("attr_price"),
+      item.getAttribute("attr_id")
+    );
+    console.log(newCL);
```
Comment (danielmargolin) [inline scripts/script.js]: console.log

File: scripts/script.js [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+    cart.addCartline(newCL);
+    // removeCartline();
```
Comment (danielmargolin) [inline scripts/script.js]: comment

File: style.css [hunk 27]
```diff
@@ -0,0 +1,356 @@
...
+    flex-direction: column;
+    min-height: 100vh;
+    max-height: 100vh;
+    overflow-x: hidden;
+}
+
+/* Header Section */
+
+header {
+    /* background: red; */
```
Comment (danielmargolin) [inline style.css]: comment

File: style.css [hunk 28]
```diff
@@ -0,0 +1,356 @@
...
+    transition: 150ms ease-in-out;
+}
+
+/* Main Section */
+
+main {
+    display: flex;
+    width: 100vw;
+    flex: 1;
+    /* background: red; */
```
Comment (danielmargolin) [inline style.css:69]: comment

File: style.css [hunk 29]
```diff
@@ -0,0 +1,356 @@
...
+main .catalog.active {
+    grid-template-columns: repeat(4, 1fr);
+}
+
+main .catalog .item {
+    border: .015rem solid black;
+    width: 100%;
+    padding: 1.5rem;
+    text-align: center;
+    /* box-shadow: .1rem .1rem .1rem; */
```
Comment (danielmargolin) [inline style.css]: comment

File: scripts/script.js [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+document.querySelector(".open-close-cart").addEventListener("click", () => {
+  document.querySelector(".cart-wrapper").classList.toggle("active");
+  document.querySelector(".catalog").classList.toggle("active");
+});
+
+document.querySelector("main").style.height =
+  window.innerHeight - document.querySelector("header").offsetHeight + "px";
+
+const catalog = new Catalog();
+data.items.map((item, i) => {
```
Comment (danielmargolin) [inline scripts/script.js:21]: maybe forEach... map will do the same effect in this case, but still...

File: scripts/script.js [hunk 31]
```diff
@@ -0,0 +1,91 @@
...
+document
+  .querySelectorAll(".catalog .item .item__img")
+  .forEach((itemImgElement) => {
+    itemImgElement.style.height = maxHeightImg + "px";
+  });
+
+console.log(maxHeightImg);
+
+const searchBar = document.querySelector(".search-bar-wrapper input");
+searchBar.addEventListener("input", () => {
```
Comment (danielmargolin) [inline scripts/script.js:38]: handleInputChange?

**Comment by danielmargolin** [comment]
where is the webpack config?

File: scripts/Item.js (hunk 2 above)
Comment (ozey5540) [reply scripts/Item.js]: you can also use backticks (``) and that way you can rid off all the + signs and use ${} for adding variables to your html string

**Pull request merged**